### PR TITLE
Gym fixes - Follow up from #705

### LINKED ISSUE #674: Gym 0.21.0 support
Hi All,
Thank you for you valuable contribution to the RL community!
I'm working with SB and [MinAtar](https://github.com/kenjyoung/MinAtar/tree/master) their latest commit requires `gym==0.21.0` which creates conflict with SB3. The current workaround is to checkout to older commit and then installing MinAtar.
When do you tend to release a version which supports `gym==0.21.0`?

Thanks in advance.

### PR DESCRIPTION
**Pull request by carlosluis**
## Description
This is a follow-up from #705, to make it merge-able and move forward

Closes #674 

## Motivation and Context
Update to new gym version, some fixes were required.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Documentation (update in the documentation)

## Checklist:

- [x] I've read the [CONTRIBUTION](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) guide (**required**)
- [x] I have updated the changelog accordingly (**required**).
- [x] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.
- [x] I have reformatted the code using `make format` (**required**)
- [x] I have checked the codestyle using `make check-codestyle` and `make lint` (**required**)
- [x] I have ensured `make pytest` and `make type` both pass. (**required**)
- [ ] I have checked that the documentation builds using `make doc` (**required**)

Note: You can run most of the checks using `make commit-checks`.

Note: we are using a maximum length of 127 characters per line

## Changelist
- Merged master
- Fixed a failing test
- Added suggestion from @vwxyzjn to include autorom in `setup.py`  (see [comment](https://github.com/DLR-RM/stable-baselines3/pull/705#discussion_r785007187))

## TODOs
- AFAIK, the remaining changes are in docs as pointed by @araffin:
> doc/tutorial need to be updated too (see comment #572 (comment)")

all of which live in other repos. Are there any remaining docs that need updating in this repo?

### CODE REVIEW

**Comment by AdamGleave** [comment]
@araffin I'm happy to review this (dependency on old Gym is causing some annoyance in another project of mine).

It seems the main things that need to be changed are in the tutorials, RL zoo, etc. Do you want to make those changes simultaneously, or merge this and then fix the other repos? If other repos are pinned to latest PyPi version it should be OK to break them on master and just fix before next release I think?

**Comment by araffin** [comment]
> @araffin I'm happy to review this (dependency on old Gym is causing some annoyance in another project of mine).
> 
> It seems the main things that need to be changed are in the tutorials, RL zoo, etc. Do you want to make those changes simultaneously, or merge this and then fix the other repos? If other repos are pinned to latest PyPi version it should be OK to break them on master and just fix before next release I think?

hey,
i'm on holidays until the 21st of february, so please go ahead (but i would keep Atari v4 for now, cf. discussion in the other pr).
and it's ok if master break other repos but only for some days (i would like the other prs to be opened even though not merged)

**Review by AdamGleave — COMMENTED** [review]
PR looks almost ready, thanks for preparing it! I've made a few mostly cosmetic suggestions for docs etc, and one more substantive suggestion in `setup.py` to simplify dependency resolution.

The other changes as flagged by araffin are:
  - We should revert the changes from v4 to v5 of the Atari environments, as v5 seems not quite ready for deployment yet per JesseFarebro's comment in https://github.com/DLR-RM/stable-baselines3/pull/572#issuecomment-999123320 and earlier discussion.
  - We should make PRs to update Zoo, tutorials, etc for new Gym version. I think this should be minor (version number changes of Pendulum etc).

Sorry this PR took a while to review @carlosluis, will you have the time to make these other changes in the next week or so? If not I'll see if I can triage it to someone else.

**Comment by AdamGleave** [comment]
I think closing this was a side-effect of deleting the branch, I'm going to reopen and try to rebase

**Comment by Miffyli** [comment]
👀 ok no more git-gymnastics for me at 3am. I have no idea how this happened...
image

**Comment by AdamGleave** [comment]
> eyes ok no more git-gymnastics for me at 3am. I have no idea how this happened... image

I think this is a GitHub bug! What I saw earlier was that you deleted `blacken` (as a result of closing the PR). I do not know why GitHub now thinks you deleted `master` :)

File: docs/guide/custom_policy.rst [hunk 1]
```diff
@@ -145,7 +145,7 @@ that derives from ``BaseFeaturesExtractor`` and then pass it to the model when t
       features_extractor_class=CustomCNN,
       features_extractor_kwargs=dict(features_dim=128),
   )
-  model = PPO("CnnPolicy", "BreakoutNoFrameskip-v4", policy_kwargs=policy_kwargs, verbose=1)
+  model = PPO("CnnPolicy", "BreakoutNoFrameskip-v5", policy_kwargs=policy_kwargs, verbose=1)
```
[the remark this answers is not in the document]
Comment (vwxyzjn) [reply docs/guide/custom_policy.rst:148]: I’d recommend don’t do this yet. I was doing some testing with the v5 via 

```python
def make_env(gym_id, seed, idx, capture_video, run_name):
    def thunk():
        env = gym.make("ALE/Breakout-v5", frameskip=1, full_action_space=False)
        env = gym.wrappers.RecordEpisodeStatistics(env)
        if capture_video:
            if idx == 0:
                env = gym.wrappers.RecordVideo(env, f"videos/{run_name}")
        env = NoopResetEnv(env, noop_max=30)
        env = MaxAndSkipEnv(env, skip=4)
        env = EpisodicLifeEnv(env)
        if "FIRE" in env.unwrapped.get_action_meanings():
            env = FireResetEnv(env)
        env = ClipRewardEnv(env)
        env = gym.wrappers.ResizeObservation(env, (84, 84))
        env = gym.wrappers.GrayScaleObservation(env)
        env = gym.wrappers.FrameStack(env, 4)
        env.seed(seed)
        env.action_space.seed(seed)
        env.observation_space.seed(seed)
        return env

    return thunk
```

And I wasn't able to reproduce past results of getting 400 scores in breakout in the v4 environments. See this [tracked experiment](https://wandb.ai/costa-huang/cleanRL/runs/ct778ly3/code?workspace=user-costa-huang) and the [related report](https://wandb.ai/costa-huang/cleanRL/reports/Breakout-v5--VmlldzoxNDI1MTIx)

File: setup.py [hunk 2]
```diff
@@ -116,7 +116,8 @@
             # For render
             "opencv-python",
             # For atari games,
-            "atari_py==0.2.6",
+            "ale-py~=0.7.1",
```
[the remark this answers is not in the document]
Comment (vwxyzjn) [reply setup.py:119]: Sure, this makes sense to me.

File: tests/test_utils.py [hunk 3]
```diff
@@ -39,11 +39,11 @@ def test_make_vec_env(env_id, n_envs, vec_env_cls, wrapper_class):
     env.close()
 
 
-@pytest.mark.parametrize("env_id", ["BreakoutNoFrameskip-v4"])
+@pytest.mark.parametrize("env_id", ["ALE/Breakout-v5"])
```
[the remark this answers is not in the document]
Comment (vwxyzjn) [reply tests/test_utils.py:42]: See previous comment. `ALE/breakout-v5` is not stable yet in my testing.

File: docs/guide/custom_policy.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (vwxyzjn) [reply docs/guide/custom_policy.rst:148]: We should probably also come up with more consistent naming and expectation. In envpool, it’s `Breakout-v5` and by default all of the Deepmind preprocessing techniques are used. In gym, it’s `ALE/Breakout-v5`, with an `ALE` prefix, and by default do not use the Deepmind preprocessing techniques and also seems to toggle sticky action by default (see [link](https://brosa.ca/blog/ale-release-v0.7))

Cc @Trinkle23897 @jkterry1 @JesseFarebro

[the remark this answers is not in the document]
Comment (AdamGleave) [reply docs/guide/custom_policy.rst:148]: Thanks for the link! Regression in performance in v5 environments could be working-as-intended, if I understand correctly part of motivation for changes is to make it harder for RL algorithms to game these environments. But big enough change that it's best deferred to another PR, once the dust settles upstream.

**Comment by carlosluis** [comment]
> Sorry this PR took a while to review @carlosluis, will you have the time to make these other changes in the next week or so? If not I'll see if I can triage it to someone else.

Hey @AdamGleave thanks for reviewing, I'm a bit swamped right now and won't have time to look at this until probably late next week or beginning of the week after. So in the interest of getting it merged, it would be great if you find someone in the mean time to finish it up!

**Review by AdamGleave — APPROVED** [review]
LGTM, will merge once tests pass

**Comment by AdamGleave** [comment]
Merging now.

@araffin I've opened PRs in SB example repos & zoo. No rush, can take a look at it after you're back from vacation :)

  - https://github.com/Stable-Baselines-Team/rl-colab-notebooks/pull/7
  - https://github.com/araffin/rl-tutorial-jnrr19/pull/12
  - https://github.com/araffin/rl-handson-rlvs21/pull/2
  - https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/pull/59

**Comment by jkterry1** [comment]
@AdamGleave why did you downgrade from the newest version of Atari environments? v5 has some extraordinarily large upgrades, see https://brosa.ca/blog/ale-release-v0.7

**Comment by AdamGleave** [comment]
> @AdamGleave why did you downgrade from the newest version of Atari environments? v5 has some extraordinarily large upgrades, see https://brosa.ca/blog/ale-release-v0.7

There is some discussion of this in https://github.com/DLR-RM/stable-baselines3/pull/734#discussion_r799055496 and araffin expressed some concerns in https://github.com/DLR-RM/stable-baselines3/pull/572#issuecomment-999094769

araffin can add his own thoughts but my desiderata before upgrading would be:
  1. Not expecting any other major upstream changes. https://github.com/DLR-RM/stable-baselines3/pull/572#issuecomment-993834080 shows there were changes happening as little as six weeks ago. People are free to use v5 environments in their code, but what we use as defaults and as test cases should be a stable target, so may make sense to wait a bit.
  2. Benchmarks to show there aren't performance regressions on these new environments, or if there are to verify this is working as intended (e.g. because sticky action makes them harder, etc).

That upgrade seems better dealt with in a new PR (which you're welcome to open).

File: setup.py [hunk 4]
```diff
@@ -116,7 +116,7 @@
             # For render
             "opencv-python",
             # For atari games,
-            "atari_py==0.2.6",
+            "gym[atari,accept-rom-license]>=0.21",
```
Comment (araffin) [inline setup.py:119]: I think this doesn't work (I remember testing it in the past), we should put autorom with accept license here

File: setup.py [hunk 5]
```diff
@@ -73,7 +73,7 @@
     packages=[package for package in find_packages() if package.startswith("stable_baselines3")],
     package_data={"stable_baselines3": ["py.typed", "version.txt"]},
     install_requires=[
-        "gym>=0.17,<0.20",  # gym 0.20 breaks atari-py behavior
+        "gym>=0.21",  # Remember to also update gym version in "extra" below when this changes
```
Comment (araffin) [inline setup.py:76]: I would probably fixed the version until https://github.com/DLR-RM/stable-baselines3/pull/780 is ready

File: setup.py (hunk 4 above)
Comment (AdamGleave) [reply setup.py:119]: I just tested it and it seems to work OK:
```
virtualenv test_venv
. ./test_venv/bin/activate
pip install -e .[extra]
```
then:
```
$ python -c 'import gym; gym.make("BreakoutNoFrameSkip-v4")'
A.L.E: Arcade Learning Environment (version 0.7.4+069f8bd)
[Powered by Stella]
```

IIRC it fails on some very old pip versions that don't support backtracking. I tested on pip 20.3.4 and Python 3.9, but I'm pretty sure it works on older versions (there's some discussion about this in the review).

**Pull request merged**